scrape/scrape.go

```diff
@@ -183,55 +193,75 @@ func newScrapePool(cfg *config.ScrapeConfig, app storage.Appendable, offsetSeed
 		escapingScheme:       escapingScheme,
 	}
 	sp.newLoop = func(opts scrapeLoopOptions) loop {
-		// Update the targets retrieval function for metadata to a new scrape cache.
-		cache := opts.cache
-		if cache == nil {
-			cache = newScrapeCache(metrics)
-		}
-		opts.target.SetMetadataStore(cache)
-
-		return newScrapeLoop(
-			ctx,
-			opts.scraper,
-			logger.With("target", opts.target),
-			buffers,
-			func(l labels.Labels) labels.Labels {
+		// NOTE: Formatting matches scrapeLoop fields order for readability.
+		sl := &scrapeLoop{
+			buffers:    buffers,
+			appendable: appendable,
+			sampleMutator: func(l labels.Labels) labels.Labels {
 				return mutateSampleLabels(l, opts.target, opts.honorLabels, opts.mrc)
 			},
-			func(l labels.Labels) labels.Labels { return mutateReportSampleLabels(l, opts.target) },
-			func(ctx context.Context) storage.Appender { return app.Appender(ctx) },
-			cache,
-			sp.symbolTable,
-			offsetSeed,
-			opts.honorTimestamps,
-			opts.trackTimestampsStaleness,
-			opts.enableCompression,
-			opts.sampleLimit,
-			opts.bucketLimit,
-			opts.maxSchema,
-			opts.labelLimits,
-			opts.interval,
-			opts.timeout,
-			opts.alwaysScrapeClassicHist,
-			opts.convertClassicHistToNHCB,
-			cfg.ScrapeNativeHistogramsEnabled(),
-			options.EnableStartTimestampZeroIngestion,
-			options.EnableTypeAndUnitLabels,
-			options.ExtraMetrics,
-			options.AppendMetadata,
-			opts.target,
-			options.PassMetadataInContext,
-			metrics,
-			options.skipOffsetting,
-			sp.validationScheme,
-			sp.escapingScheme,
-			opts.fallbackScrapeProtocol,
-		)
+			reportSampleMutator: func(l labels.Labels) labels.Labels { return mutateReportSampleLabels(l, opts.target) },
+			offsetSeed:          offsetSeed,
+			metrics:             metrics,
+
+			symbolTable:      sp.symbolTable,
+			validationScheme: sp.validationScheme,
+			escapingScheme:   sp.escapingScheme,
+
+			enableNativeHistogramScraping: cfg.ScrapeNativeHistogramsEnabled(),
+
+			enableSTZeroIngestion:   options.EnableStartTimestampZeroIngestion,
+			enableTypeAndUnitLabels: options.EnableTypeAndUnitLabels,
+			reportExtraMetrics:      options.ExtraMetrics,
+			appendMetadataToWAL:     options.AppendMetadata,
+			skipOffsetting:          options.skipOffsetting,
+
+			scrapeLoopOptions: opts,
+		}
+		sl.init(ctx, options.PassMetadataInContext)
+		return sl
 	}
 	sp.metrics.targetScrapePoolTargetLimit.WithLabelValues(sp.config.JobName).Set(float64(sp.config.TargetLimit))
 	return sp, nil
 }
 
+// init prepares scrapeLoop after raw construction.
+// NOTE: While newScrapeLoop constructor pattern would be safer, it has proven to be
+// highly not readable (too many params). Instead, we follow init pattern.
```

Comment:
I think this commentary belongs in the commit message not the code.
It's talking about history rather than something relevant when you are reading the code.